### ISSUE TITLE
Update pydantic to v2 API

### ISSUE THREAD

**Issue by matthewfeickert** [ssl-hep/ServiceX_frontend#363]
At the moment on the `3.0_develop` branch, the `pydantic` requirement is constrained to

https://github.com/ssl-hep/ServiceX_frontend/blob/a7f11f50e0074267fdb9f8e35d85b12103d2a97f/pyproject.toml#L38)

(as that branch isn't stable forever, for posterity

```toml
dependencies = [
    "qastle>=0.17",
    "func_adl>=3.2.6",
    "requests>=2.31",
    "pydantic>=1.10,<2.0",  # FIXME: Adopt pydantic 2.0 API
    ...
]
```

)

In https://github.com/iris-hep/idap-200gbps-atlas/issues/41 @gordonwatts had identified this as a blocker for using in the same stack as other tools like `correctionlib`. Updating `pydantic` to use `v2` would be highly impactful.

**Comment by matthewfeickert** [comment]
cc @kyungeonchoi and @BenGalewsky as they are probably the relevant people to discuss this with beyond / in addition to @gordonwatts.

**Comment by matthewfeickert** [comment]
I tried to give this a go on my fork's `feat/update-to-pydantic-v2-api` branch, but didn't get super far as I'm inexperienced with `pydantic` and haven't yet worked through the [SciPy 2023 `pydantic` tutorial](https://github.com/adonath/scipy-2023-pydantic-tutorial) to teach myself. Example, it took me over 30 minutes to understand that

https://github.com/ssl-hep/ServiceX_frontend/blob/a7f11f50e0074267fdb9f8e35d85b12103d2a97f/servicex/databinder_models.py#L121

(Posterity repost as the `3.0_develop` branch will not live forever
```python
class ServiceXSpec(BaseModel):
    General: General
    Sample: List[Sample]
    Definition: Optional[Definition]
```
)

is getting hit by the field having the same name as a model (https://github.com/pydantic/pydantic/issues/7871#issuecomment-1770958931) and so you need to change the field name

```python
...
    definition: Optional[Definition]
```

.

@BenGalewsky as you were the one to move things over to `pydantic` in PR https://github.com/ssl-hep/ServiceX_frontend/pull/340 you might be in the best position to either give advice if you have the time to scout the differences, or also do the implementation changes if they all appear trivial to you.

**Comment by matthewfeickert** [comment]
Resolved by PR #368
